### ISSUE TITLE
Computing gradients with Tensorflow 2.4.1: 'KerasTensor' object has no attribute '_id'

### ISSUE THREAD

**Issue by jwallbridge** [tensorflow/tensorflow#48531]
I am trying to manipulate gradients within a layer in Tensorflow 2.4.1.  There seems to be a problem leading to the error 'KerasTensor' object has no attribute '_id'.  I am using a custom loss function.

I have linked standalone code below that reproduces the issue.  This is the bare minimum necessary to generate the problem.  My actual working code is part of a large RL pipeline so I am looking for a solution without modifying the general structure of the linked example.  I suspect it is a bug.

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes.
- TensorFlow installed from (source or binary): Colab
- TensorFlow version (use command below): 2.4.1
- Python version: 3.7.10

**Describe the current behavior**

AttributeError: 'KerasTensor' object has no attribute '_id' when computing gradients using custom loss function in a lambda layer.

**Describe the expected behavior**

The model to compile without error.

**Standalone code to reproduce the issue**

https://drive.google.com/file/d/11jAwZM2cq-4ng9Spo03dWxg_H5iaMbVN/view?usp=sharing

Thank you in advance for the assistance.

**Comment by AdityaKane2001** [comment]
@jwallbridge 
The correct usage for gradient tape is:
```python
with tf.GradientTape() as tape:
		# make a prediction using the model and then calculate the
		# loss
		pred = model(X)
		loss = custom_loss_fn(y, pred)
	# calculate the gradients using our tape and then update the
	# model weights
	grads = tape.gradient(loss, model.trainable_variables)
	opt.apply_gradients(zip(grads, model.trainable_variables))
```

Also, you need not make a Lambda layer for the loss function. The loss function `customLoss` is good enough. You can pass `loss = customLoss` in the model.

**Comment by tilakrayal** [comment]
@jwallbridge,

Please take a look at @AdityaKane2001's comment and let us know if you are still facing the same issue?  Thanks!

**Comment by jwallbridge** [comment]
Thank you @AdityaKane2001 and @tilakrayal.  I am still facing the same issue.  The problem with @AdityaKane2001's comment is the `X` in `model(X)`.  Let's say I want to rescale the gradients in a layer by 1/3.  In Tensorflow 1 you could simply do 
```python
optimizer = Adam(learning_rate=0.01)
var_list = model.get_layer(name='mid_layer').trainable_variables
gradients = optimizer.compute_gradients(objective, var_list=var_list)
for i, (grad, var) in enumerate(gradients):
    if grad is not None:
        scale_val = 1.0 / 3
        grad *= scale_val
optimizer.apply_gradients(gradients)
```
without an explicit input.  It seems `compute_gradients` is not available in Tensorflow 2.4.1.  When you replace `model(X)` by `model.output` you get 
```
TypeError: Cannot convert function customLoss to a TensorFlow DType.
```
This is due to `pred=model.output` in GradientTape.  

There must be a simple way to do this in Tensorflow 2.4.1?  

For more context, I would like to scale particular layer gradients within an RL pipeline consisting of a compile module (associated to an abstract base class) of the form
```python
def compile(self, optimizer, metrics=[]):
    ...
    y_pred = self.model.output
    y_true = Input(name='y_true', shape=(self.nb_actions,))
    ....
    SCALE GRADIENTS HERE
    var_list = model.get_layer(name='mid_layer').trainable_variables
    optimizer.apply_gradients(zip(grads, var_list))
    ....
    model.compile(optimizer=optimizer, loss=customLoss, metrics=combined_metrics)
````
As you can see, there is no explicit `X`.

**Comment by AdityaKane2001** [comment]
@jwallbridge 
I am a bit confused what you are trying to do here. If I'm not mistaken, you want to write a custom training loop which will scale the gradients by some constant and then apply them, right? Or is it something else that you are trying to achieve?

**Comment by jwallbridge** [comment]
@AdityaKane2001 
Thank you for your assistance.  Indeed, I would like to scale (and clip) gradients in a chosen layer of my network.  I am following the official documentation
https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer
which states that the correct procedure is
```python
# Create an optimizer.
opt = tf.keras.optimizers.SGD(learning_rate=0.1)

# Compute the gradients for a list of variables.
with tf.GradientTape() as tape:
  loss = <call_loss_function>
vars = <list_of_variables>
grads = tape.gradient(loss, vars)

# Process the gradients, for example cap them, etc.
# capped_grads = [MyCapper(g) for g in grads]
processed_grads = [process_gradient(g) for g in grads]

# Ask the optimizer to apply the processed gradients.
opt.apply_gradients(zip(processed_grads, var_list))
```
contrary to your suggestion of including `pred = model(X)`.  When you follow these instructions you get my original error which is why I considered this a bug.  

For complete context, I am modifying the compile method of existing code from the well known Keras-RL:
https://github.com/keras-rl/keras-rl/blob/master/rl/agents/dqn.py#L167
The training loop with a fit method already exists.  I just want to manipulate gradients within the compile method.  With the old `compute_gradients` this is simple.  So what replaces `compute_gradients` in TF 2.4.1 or could this be a bug with GradientTape in this context?

**Comment by tilakrayal** [comment]
@jvishnuvardhan ,
I was able to reproduce the issue in TF v2.4,v2.5.0rc0,nightly.Please find the [gist](https://colab.research.google.com/gist/tilakrayal/7b6386e1bf710ea29e5e1f9e4657c0f3/48531.ipynb) here

**Comment by jwallbridge** [comment]
Thank you @tilakrayal for reproducing the issue.
@jvishnuvardhan ,
do you have an update on the issue?

**Comment by jvishnuvardhan** [comment]
@jwallbridge Can you please check this [issue](https://github.com/tensorflow/tensorflow/issues/46194#issuecomment-827036146) which is similar to your issue. Thanks!

**Comment by code2k13** [comment]
This issue still persists. In my example, I have a custom loss function that uses mae over intermediate layers of a neural network. The `model.get_layer("xxx").output` API returns a 'KerasTensor' which I believe causes this issue.

**Comment by AdityaKane2001** [comment]
> API returns a 'KerasTensor' which I believe causes this issue.

Yes, and as per the [source code](https://github.com/keras-team/keras/blob/3613c3defc39c236fb1592c4f7ba1a9cc887343a/keras/engine/keras_tensor.py), KerasTensor is in no way related to tf.Tensor. It just overloads all methods of tf.Tensor, so it is interchangably usable in operations (I guess this is the motive of doing the overloading). 

I am not aware of any method to convert a KerasTensor to a tf.Tensor. As mentioned in the [docstring](https://github.com/keras-team/keras/blob/3613c3defc39c236fb1592c4f7ba1a9cc887343a/keras/engine/keras_tensor.py#L32-L111), KerasTensor has limitations when it comes to functional API. 

Maybe this issue is best opened in Keras' repo itself. Thus, I'll be doing the same soon.

/cc @haifeng-jin @fchollet

**Comment by code2k13** [comment]
Thanks, but how does one create a loss function that needs output from intermediate layers. Is there any other way? Would be really grateful if someone can guide me.
